openwrt/luci
# luci-app-example: Update with more documentation, more examples

Summary
=======

The existing luci-app-example is very basic, and provides very little context to someone just getting started with LuCI development.

The changes in this PR are basically "I wish I'd had this information and context available to me", as I've recently worked on a PR to provide a UI for AdGuard Home, and it took me several days of hunting around in the wiki, other LuCI apps, and general 'net searches.

This PR:

* adds a structure.md file to explain the structure of a LuCI app
* updates README.md with some details about the JS files in the example
* updates README.md with a note about ACLs
* updates the ACL file based on testing, removing unused grants
* updates the menu file with links to the new JS UI files
* adds RPC examples for a dedicated ubus RPC endpoint that's not UCI
* adds a Lua script to answer RPC requests via rpcd

The Lua script demonstrates a few techniques, such as

* converting Lua tables to JSON
* converting YAML files to JSON for UI consumption
* argument parsing, with explanation about why it's done this way (luci.ddns did not explain the logic, nor did the wiki)
* how to test the Lua script without ubus
* how to test the Lua script with ubus

Three variants of the rpc.js file are provided. The first one is just calling RPC endpoints and rendering what could be equated to a status page; editing is turned off. The second and third call the same RPC endpoint, and demonstrate two different layout approaches - TableSection vs TypedSection.

Testing
=======

All testing has been manual.

ACLs were refined by moving acl.d/luci-compat.json out of the way to see what broke in the UI, and then tweaking the ACLs until the UI worked again.

Arg parsing for the Lua script is based on the luci.ddns script, and it feels like the parseInput and validateArgs functions should be a core library so that Lua scripts for rpcd have a unified way of working.

Signed-off-by: Duncan Hill <openwrt-dev@cricalix.net>
